### ISSUE TITLE
documentbuilderfactory-disallow-doctype-decl-missing.java doesn't work with static variables

### ISSUE THREAD

**Issue by coheigea** [returntocorp/sgrep-rules#2496]
**Describe the bug**
documentbuilderfactory-disallow-doctype-decl-missing.java does not work with java static variables

**To Reproduce**
```
class GoodDocumentBuilderFactoryStatic {
    private static DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();

    static {
        //ok:documentbuilderfactory-disallow-doctype-decl-missing
        dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
    }
}
```

**Expected behavior**
The rule should be able to match static blocks with static variables and see that doctypes are correctly disabled. It looks like the current rule only matches inside a method. If you give me some pointers about how I might do this, then I'll test + prepare a PR.

**Priority**
How important is this to you?
- [ ] P2: annoying but not blocking me

**Comment by LewisArdern** [comment]
cc @inkz @0xDC0DE

**Comment by 0xDC0DE** [comment]
Hi @coheigea, I actually reported a related issue to our Program Analysis team yesterday. This is something that we can probably work around in the rules if we convert them to taint rules.

The source can be a new parser (or in this can DocumentBuilderFactory), the sink can be wherever we use the parser, and the configurations are the sanitizers. But there are a lot of rules that require being rewritten, and this is a common pattern. The pattern that I ran into was this:

```
DocumentBuilderFactory dbf = null;
if (condition) {
    dbf = ...;
else {
   dbf = ...;
}
dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
```

I would hold off on rewriting this rule, because the fix might be coming in semgrep matching instead.

**Comment by 0xDC0DE** [comment]
We decided to rewrite rules like this with taint mode.
However, currently, taint mode only looks at the current function and not at class attributes (yet). If we want to taint the sink (`dbf.newDocumentBuilder()`) when the factory is a static class atttribute, that is currently not possible. So rewriting  this rule to taint does not solve your issue: https://semgrep.dev/playground/s/Dl1y

Also, we currently can not match static blocks: https://semgrep.dev/playground/s/ojQ0
It's a known issue that hasn't gotten any high priority at the moment. So updating this rule or the taint version rule to get around your use-case specifically will also be difficult. 

In conclusion, we'll have to wait for Semgrep engine updates before your specific false positive can be fixed. But these types of rules should be updated to solve a few other related false positives.

**Comment by coheigea** [comment]
Thanks for the update!

**Comment by 0xDC0DE** [comment]
Another update. Looks like we found a workaround to match on a `static` block. It works as long as we add a class around it. This is required to find an entry point for the parser.

https://semgrep.dev/s/2vGj

So we can update the rule! You mentioned you wanted to test and prepare a PR, you can still if you want to. If you need help find me in the community slack as Pieter De Cremer! If you don't get around to it, I will pick this up somewhere next week probably.
